doc/mgr/dashboard: Add troubleshooting guide (#34855)
ceph/ceph: doc/mgr/dashboard.rst

.. include:: dashboard_plugins/feature_toggles.inc.rst
.. include:: dashboard_plugins/debug.inc.rst

Troubleshooting the Dashboard
-----------------------------

Locating the Dashboard
^^^^^^^^^^^^^^^^^^^^^^

If you are unsure of the location of the Ceph Dashboard, run the following command::

    $ ceph mgr services | jq .dashboard
    "https://host:port"

The command returns the URL where the Ceph Dashboard is located: ``https://<host>:<port>/``

Accessing the Dashboard
^^^^^^^^^^^^^^^^^^^^^^^

If you are unable to access the Ceph Dashboard, run through the following
commands:

#. Verify the Ceph Dashboard module is enabled::

    $ ceph mgr module ls | jq .enabled_modules

   Ensure the Ceph Dashboard module is listed in the return value of the
   command. Example snipped output from the command above::

    [
      "dashboard",
      "iostat",
      "restful"
    ]

#. If it is not listed, activate the module with the following command::

    $ ceph mgr module enable dashboard

#. Check the Ceph Dashboard and/or mgr log file for any errors. The exact
   location of the log files depends on the Ceph configuration.

>    * Check if the mgr log message are written to a file by::

Reviewer: ```suggestion
   * Check if mgr log messages are written to a file by::
```

        $ ceph config get mgr log_to_file
        true

   * Get the location of the log file (it's ``/var/log/ceph/<cluster-name>-<daemon-name>.log``
     by default)::

        $ ceph config get mgr log_file
        /var/log/ceph/$cluster-$name.log

#. Ensure the SSL/TSL support is configured properly:

   * Check if the SSL/TSL support is enabled::

       $ ceph config get mgr mgr/dashboard/ssl

   * If the command returns ``true``, verify a certificate exists by::

       $ ceph config-key get mgr/dashboard/crt

     and::

       $ ceph config-key get mgr/dashboard/key

   * If it doesn't, run the following command to generate a self-signed

>      certificate::

Reviewer: How about this?
```suggestion
     certificate or follow the instructions outlined in :ref:`dashboard-ssl-tls-support`::
     on how to import custom SSL certificates::
```

    $ Error ENOENT: User <username> does not exist

#. Check if the user is enabled::

    $ ceph dashboard ac-user-show <username> | jq .enabled
    true

   Check if ``enabled`` is set to ``true`` for your user. If not the user is
   not enabled, run::

    $ ceph dashboard ac-user-enable <username>

A Dashboard Feature is Not Working
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

When an error occurs on the backend, you will usually receive an error
notification on the frontend. Run through the following scenarios to debug.

#. Check the Ceph Dashboard/mgr logfile(s) for any errors. These can be
   identified by searching for keywords, such as *500 Internal Server Error*,
   followed by ``traceback``. The end of a traceback contains more details about
   what exact error occurred.
#. Check your web browser's Javascript Console for any errors.

Ceph Dashboard Logs
^^^^^^^^^^^^^^^^^^^

Dashboard Debug Flag
''''''''''''''''''''

With this flag enabled, traceback of errors are included in backend responses.

To enable this flag via the Ceph Dashboard, navigate from *Cluster* to *Manager
modules*. Select *Dashboard module* and click the edit button. Click the
*debug* checkbox and update.

> To enable via the CLI, run the following command::

Reviewer 2: ```suggestion
To enable it via the CLI, run the following command::
```

.. include:: dashboard_plugins/feature_toggles.inc.rst
.. include:: dashboard_plugins/debug.inc.rst

Troubleshooting the Dashboard
-----------------------------

Locating the Dashboard
^^^^^^^^^^^^^^^^^^^^^^

If you are unsure of the location of the Ceph Dashboard, run the following command::

    $ ceph mgr services | jq .dashboard
    "https://host:port"

The command returns the URL where the Ceph Dashboard is located: ``https://<host>:<port>/``

> 

Reviewer: With regards to the comment from @SchoolGuy : maybe we should add a hint about installing `jq` here?

```
Note that many Ceph command line tools return results in JSON format.
You may have to install the ``jq`` command-line JSON processor utility
on your operating system beforehand.
```

Author: Thanks for the text @LenzGr! My idea would be to add a link to the jq website as well:

```
Note: Many Ceph command line tools return results in JSON format.
You may have to install the `jq <https://stedolan.github.io/jq>`_ 
command-line JSON processor utility on your operating system beforehand.
```

Reviewer 3: @s0nea That would indeed be the most superb solution in my eyes!

       $ ceph dashboard create-self-signed-cert

Trouble Logging into the Dashboard
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you are unable to log into the Ceph Dashboard and you receive the following
error, run through the procedural checks below:

.. image:: ../images/dashboard/invalid-credentials.png
   :align: center

#. Check that your user credentials are correct. If you are seeing the
   notification message above when trying to log into the Ceph Dashboard, it
   is likely you are using the wrong credentials. Double check your username
   and password, and ensure the caps lock key is not enabled by accident.

#. If your user credentials are correct, but you are experiencing the same
   error, check that the user account exists::

    $ ceph dashboard ac-user-show <username>

   This command returns your user data. If the user does not exist, it will
   print::

    $ Error ENOENT: User <username> does not exist

#. Check if the user is enabled::

    $ ceph dashboard ac-user-show <username> | jq .enabled
    true

   Check if ``enabled`` is set to ``true`` for your user. If not the user is
   not enabled, run::

    $ ceph dashboard ac-user-enable <username>

> 

Reviewer 3: Maybe add here a link to: https://docs.ceph.com/docs/master/mgr/dashboard/#user-and-role-management

       $ ceph config get mgr mgr/dashboard/ssl

   * If the command returns ``true``, verify a certificate exists::

       $ License file check

   * If not, run to generate a self-signed certificate::

       $ ceph dashboard create-self-signed-cert

Trouble Logging into the Ceph Dashboard
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you are unable to log into the Ceph Dashboard and you receive the following
error, run through the procedural checks below:

.. image:: images/dashboard/invalid-credentials.png

#. Check that your user credentials are correct. If you are seeing the
   notification message above when trying to log into the Ceph Dashboard, it
   is likely you are using the wrong credentials. Double check your username
   and password, and ensure the caps lock key is not enabled by accident.

#. If your user credentials are correct, but you are experiencing the same
   error, check that the user account exists::

    $ ceph dashboard ac-user-show <username>

   This command returns your user data. If the user does not exist, it will
   print::

    $ Error ENOENT: User <username> does not exist

#. Check if the user is enabled::

>     $ ceph dashboard ac-user-show <username> {..., "enabled": true, ...}

Reviewer 2: I don't know how, but i think we should mark or mention that `{..., "enabled": true, ...}` is the output of the given command. Currently it looks more that this is part of the command line/arguments.

Author: Thanks for the comment. It was supposed to be in two different lines. I forgot the line break.

   followed by ``traceback``. The end of a traceback contains more details about
   what exact error occurred.
#. Check your web browser's Javascript Console for any errors.

Ceph Dashboard Logs
^^^^^^^^^^^^^^^^^^^

Debug Ceph Dashboard Flag
'''''''''''''''''''''''''

With this flag enabled, traceback of errors are included in backend responses.

To enable this flag via the Ceph Dashboard, navigate from *Cluster* to *Manager
modules*. Select *Dashboard module* and click the edit button. Click the
*debug* checkbox and update.

To enable via the CLI, run the following command::

    $ ceph dashboard debug enable

Setting Logging Level of Ceph Dashboard Module
''''''''''''''''''''''''''''''''''''''''''''''

Setting the logging level to debug makes the log more verbose and helpful for
debugging.

#. Increase the logging level of manager daemons::

   $ ceph tell mgr config set debug_mgr 20

#. Adjust the logging level of the Ceph Dashboard module via the Ceph Dashboard
   or CLI:

   * Navigate from *Cluster* to *Manager modules*. Select *Dashboard module*
     and click the edit button. Modify the ``log_level`` configuration.

>    * To adjust via the CLI, run the following command::

Reviewer 2: ```suggestion
   * To adjust it via the CLI, run the following command::
```

.. include:: dashboard_plugins/feature_toggles.inc.rst
.. include:: dashboard_plugins/debug.inc.rst

> Troubleshooting the Ceph Dashboard

Reviewer: Since this section is now part of the Ceph Dashboard chapter, we can probably trim this to:
```suggestion
Troubleshooting
```

-----------------------------

Locating the Dashboard
^^^^^^^^^^^^^^^^^^^^^^

If you are unsure of the location of the Ceph Dashboard, run the following command::

>     $ ceph mgr services | grep dashboard
>     "dashboard": "https://host:port"

Reviewer: Instead of `grep`, how about using `jq` here?
```suggestion
    $ ceph mgr services | jq .dashboard
    "https://host:port"
```

The command returns the URL where the Ceph Dashboard is located: ``https://<host>:<port>/``

Accessing the Dashboard
^^^^^^^^^^^^^^^^^^^^^^^

If you are unable to access the Ceph Dashboard, run through the following
commands:

#. Verify the Ceph Dashboard module is enabled::

    $ ceph mgr module ls

#. Ensure the Ceph Dashboard module is listed in the ``enabled_modules``

>    section. Example snipped output from the ``ceph mgr module ls`` command::

Reviewer: This is a great example to use `jq` again:
```suggestion
   section. Example snipped output from the ``ceph mgr module ls | jq .enabled_modules`` command::
```
It would return:
```
[
  "dashboard",
  "iostat",
  "restful"
]
```

        "progress",
        "rbd_support",
        "status",
        "telemetry",
        "volumes"
      ],
      "enabled_modules": [
        "dashboard",
        "iostat",
        "restful"
      ],
      "disabled_modules": [
        {
          ...
        }
      ]
    }

#. If it is not listed, activate the module with the following command::

    $ ceph mgr module enable dashboard

#. Check the Ceph Dashboard and/or mgr log file for any errors. The exact
   location of the log files depends on the Ceph configuration.

   * Check if the mgr log message are written to a file by::

        $ ceph config get mgr log_to_file
        true

   * Get the location of the log file (it's ``/var/log/ceph/<cluster-name>-<daemon-name>.log``
     by default)::

        $ ceph config get mgr log_file
        /var/log/ceph/$cluster-$name.log

#. Ensure the SSL/TSL support is configured properly:

   * Check if the SSL/TSL support is enabled::

       $ ceph config get mgr mgr/dashboard/ssl

   * If the command returns ``true``, verify a certificate exists by::

       $ ceph config-key get mgr/dashboard/crt

     and::

       $ ceph config-key get mgr/dashboard/key

>    * If not, run to generate a self-signed certificate::

Reviewer: ```suggestion
   * If it doesn't, run the following command to generate a self-signed certificate::
```

       $ ceph dashboard create-self-signed-cert

Trouble Logging into the Dashboard
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you are unable to log into the Ceph Dashboard and you receive the following
error, run through the procedural checks below:

.. image:: ../images/dashboard/invalid-credentials.png
   :align: center

#. Check that your user credentials are correct. If you are seeing the
   notification message above when trying to log into the Ceph Dashboard, it
   is likely you are using the wrong credentials. Double check your username
   and password, and ensure the caps lock key is not enabled by accident.

#. If your user credentials are correct, but you are experiencing the same
   error, check that the user account exists::

    $ ceph dashboard ac-user-show <username>

   This command returns your user data. If the user does not exist, it will
   print::

    $ Error ENOENT: User <username> does not exist

#. Check if the user is enabled::

>     $ ceph dashboard ac-user-show <username>

Reviewer: ```suggestion
    $ ceph dashboard ac-user-show <username> | jq .enabled
    true
```
